scripts/sell_unused_slots.py

```diff
@@ -0,0 +1,112 @@
+log = logging.getLogger(__name__)
+
+
+class ReindexMonitor:
+    # Minutes
+    interval = 5
+
+    # Minimum number of `-contribute` Lambda invocations per interval for a
+    # reindexing to be considered active
+    threshold = 0
+
+    @cached_property
+    def _cloudwatch(self):
+        return aws.client('cloudwatch')
+
+    @cached_property
+    def _lambda(self):
+        return aws.client('lambda')
+
+    def is_reindex_active(self) -> bool:
+        for function_name in self.list_contribution_lambdas():
+            invocation_count = self._lambda_invocation_count(function_name)
+            logging_args = (function_name, f'{invocation_count} invocations in the last '
+                                           f'{self.interval} minutes')
+            if invocation_count > self.threshold:
+                log.info('Detected active reindex for Lambda %r; %s', *logging_args)
+                return True
+            else:
+                log.debug('Lambda %r is inactive (%s)', *logging_args)
+        return False
+
+    def list_contribution_lambdas(self) -> Iterable[str]:
+        """
+        Search Lambda functions for the names of contribution Lambdas.
+
+        This method retrieves over 350 kilobytes of data from AWS if its output
+        is greedily consumed.
+        """
+        # The default and maximum number of items returned is 50.
+        # Reducing this number will result in more round trips being made
+        # here when no reindex is active, but will reduce the cumulative
+        # response size if an active reindex is found early in the loop.
+        kwargs = dict(MaxItems=20)
+        while True:
+            response = self._lambda.list_functions(**kwargs)
+            for function in response['Functions']:
+                function_name = function['FunctionName']
+                if function_name.endswith('-contribute'):
+                    yield function_name
+            next_marker = response.get('NextMarker')
+            if next_marker is None:
+                break
+            else:
+                kwargs['Marker'] = next_marker
+
+    def _lambda_invocation_count(self, function_name: str) -> int:
+        now = datetime.utcnow()
+        then = now - timedelta(minutes=self.interval)
+        response = self._cloudwatch.get_metric_statistics(
+            Namespace='AWS/Lambda',
+            MetricName='Invocations',
+            Dimensions=[{
+                'Name': 'FunctionName',
+                'Value': function_name,
+            }],
+            StartTime=then,
+            EndTime=now,
+            Period=60 * self.interval,
+            Statistics=['Sum'],
+        )
+        # Expecting either zero or one data points depending on whether
+        # there were any invocations or not.
+        return sum(interval['Sum'] for interval in response['Datapoints'])
+
+
+def main(argv):
+    slot_manager = SlotManager()
+    resources = slot_manager.get_resource_names()
+    if resources is None:
```

Comment:
There is no need to leak the "resource name" concept to clients of the slot manager.

```suggestion
    if slot_manager.has_active_slots():
```